compiler/dyno/lib/uast/chpl-syntax-printer.cpp

```diff
@@ -0,0 +1,1065 @@
+    }
+
+    void visit(const Coforall* node) {
+      ss_ << "coforall ";
+
+      if (node->index()) {
+        printAst(ss_, node->index());
+        ss_ << " in ";
+      }
+      printAst(ss_, node->iterand());
+      if (node->withClause()) {
+        ss_<< " ";
+        printAst(ss_, node->withClause());
+      }
+      ss_ << " ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const Comment* node){
+      // TODO: Discuss if we want to provide comments back
+      ss_ << node->c_str();
+    }
+
+    void visit(const Conditional* node) {
+      ss_ << "if ";
+      printAst(ss_, node->condition());
+      ss_ << " ";
+      printBlockWithStyle(ss_, node->thenBlockStyle(), node->thenStmts(), "then ");
+      if (node->hasElseBlock()) {
+        ss_ << " else ";
+        printBlockWithStyle(ss_, node->elseBlockStyle(), node->elseStmts());
+      }
+    }
+
+    void visit(const CStringLiteral* node) {
+      ss_ << "c\"" << quoteStringForC(std::string(node->str().c_str())) << '"';
+    }
+
+    void visit(const Defer* node) {
+      ss_ << "defer ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts());
+    }
+
+    void visit(const Delete* node) {
+      ss_ << "delete ";
+      interpose(ss_, node->exprs(), ", ");
+    }
+
+    void visit(const Domain* node) {
+      if (node->numExprs() == 1 && node->expr(0)->isTypeQuery()) {
+        printAst(ss_, node->expr(0)->toTypeQuery());
+      } else if (node->numExprs() == 0) {
+        // do nothing
+      } else {
+        interpose(ss_, node->exprs(), ", ", "{", "}");
+      }
+    }
+
+    void visit(const Dot* node) {
+      printAst(ss_, node->receiver());
+      ss_ << "." << node->field().c_str();
+    }
+
+    void visit(const DoWhile* node) {
+      ss_ << "do ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts());
+      ss_ << " while ";
+      printAst(ss_, node->condition());
+    }
+
+    void visit(const Enum* node) {
+      ss_ << "enum " << node->name().c_str() << " ";
+      interpose(ss_, node->enumElements(), ", ", "{ ", " }");
+    }
+
+    void visit(const EnumElement* node) {
+      ss_ << node->name().c_str();
+      if (const Expression* ie = node->initExpression()) {
+        ss_ <<  " = ";
+        printAst(ss_, ie);
+      }
+    }
+
+    //erroneousExpression
+
+    void visit(const ExternBlock* node) {
+      ss_ << "extern {\n";
+      ss_ << node->code();
+      ss_ << "}";
+    }
+
+    void visit(const FnCall* node) {
+      const Expression* callee = node->calledExpression();
+      if (!callee) {
+        printf("ERROR %s\n", asttags::tagToString(node->tag()));
+        node->stringify(std::cerr, StringifyKind::CHPL_SYNTAX);
+      }
+      assert(callee);  // This should be true because OpCall is handled
+
+      printAst(ss_, callee);
+      if (callee->isIdentifier() && (callee->toIdentifier()->name() == "borrowed" || callee->toIdentifier()->name() == "owned"
```

Comment:
Let's make a helper method to handle these special cases, it can return `true` if it did the special printing.